### PR TITLE
feat(run): run logging data list by requester API

### PR DESCRIPTION
**Pull request by joremysh**
Because

- the data that users could see on dashboard are different from what they can see on Runs page in a single pipeline or model

This commit

- add new run logging API for dashboard

image

### CODE REVIEW

**Review by jvallesm — COMMENTED** [review]
I added some comments on [this PR](https://github.com/instill-ai/protobufs/pull/472), but I see you used another one for the contracts. Still, some of these comments are relevant, so please review them 🙏 . Once we agree on these, the draft PR can be closed.

File: pkg/service/pipeline.go [hunk 1]
```diff
@@ -1988,3 +1988,72 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+
+	log, _ := logger.GetZapLogger(ctx)
+
+	orderBy, err := ordering.ParseOrderBy(req)
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now().UTC()
+	startedTimeBegin := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
```
[the remark this answers is not in the document]
Comment (joremysh) [reply pkg/service/pipeline.go:2006]: They have at here: https://github.com/instill-ai/protobufs/blob/main/vdp/pipeline/v1beta/pipeline.proto#L1900
Using the same default value as metric APIs here: https://github.com/instill-ai/protobufs/blob/main/core/mgmt/v1beta/metric.proto#L159

[the remark this answers is not in the document]
Comment (jvallesm) [reply pkg/service/pipeline.go:2006]: 👍 sorry, I'd missed this

File: pkg/service/component_definition.go [hunk 2]
```diff
@@ -13,10 +13,12 @@ import (
 
 	"github.com/instill-ai/pipeline-backend/pkg/recipe"
 	"github.com/instill-ai/pipeline-backend/pkg/repository"
-	"github.com/instill-ai/x/paginate"
 
 	component "github.com/instill-ai/pipeline-backend/pkg/component/store"
 	errdomain "github.com/instill-ai/pipeline-backend/pkg/errors"
+
+	"github.com/instill-ai/x/paginate"
+
```
[the remark this answers is not in the document]
Comment (joremysh) [reply pkg/service/component_definition.go:22]: fixed

File: pkg/service/pipeline.go [hunk 3]
```diff
@@ -1988,3 +1988,72 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+	if req.GetStart().IsValid() {
+		startedTimeBegin = req.GetStart().AsTime()
+	}
+	startedTimeEnd := now
+	if req.GetStop().IsValid() {
+		startedTimeEnd = req.GetStop().AsTime()
+	}
+
+	if startedTimeBegin.After(startedTimeEnd) {
```
[the remark this answers is not in the document]
Comment (joremysh) [reply pkg/service/pipeline.go:2018]: updated

File: pkg/service/pipeline.go [hunk 4]
```diff
@@ -1988,3 +1988,72 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+		return nil, fmt.Errorf("time range end time %s is earlier than start time %s", startedTimeEnd.Format(time.RFC3339), startedTimeBegin.Format(time.RFC3339))
+	}
+	log.Info("ListPipelineRunsByCreditOwner", zap.Time("startedTimeBegin", startedTimeBegin), zap.Time("startedTimeEnd", startedTimeEnd))
```
[the remark this answers is not in the document]
Comment (joremysh) [reply pkg/service/pipeline.go:2021]: removed

[the remark this answers is not in the document]
Comment (joremysh) [reply pkg/service/pipeline.go]: rebased main to fix conflicts and resolved all comments

File: pkg/handler/pipeline.go [hunk 5]
```diff
@@ -1996,3 +1996,37 @@ func (h *PublicHandler) ListComponentRuns(ctx context.Context, req *pb.ListCompo
 
 	return resp, nil
 }
+
+func (h *PublicHandler) ListPipelineRunsByCreditOwner(ctx context.Context, req *pb.ListPipelineRunsByCreditOwnerRequest) (*pb.ListPipelineRunsByCreditOwnerResponse, error) {
```
[the remark this answers is not in the document]
Comment (joremysh) [reply pkg/handler/pipeline.go:1999]: updated

File: pkg/service/pipeline.go [hunk 6]
```diff
@@ -1993,3 +1991,84 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+	if req.GetStart().IsValid() {
+		startedTimeBegin = req.GetStart().AsTime()
+	}
+	startedTimeEnd := now
+	if req.GetStop().IsValid() {
+		startedTimeEnd = req.GetStop().AsTime()
+	}
+
+	if startedTimeBegin.After(startedTimeEnd) {
+		return nil, fmt.Errorf("time range end time %s is earlier than start time %s", startedTimeEnd.Format(time.RFC3339), startedTimeBegin.Format(time.RFC3339))
```
Comment (jvallesm) [inline pkg/service/pipeline.go:2032]: nit: I'd keep the error message independent from the actual time values. That might potentially help us group the error message and evaluate the severity of a bug.

File: pkg/service/pipeline.go [hunk 7]
```diff
@@ -1993,3 +1991,84 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+	}
+
+	pipelineRuns, totalCount, err := s.repository.GetPaginatedPipelineRunsByRequester(ctx, requesterUID, startedTimeBegin, startedTimeEnd,
```
Comment (jvallesm) [inline pkg/service/pipeline.go:2035]: nit: with more than 3 arguments, I'd consider passing a request / params struct rather than a long list of args

File: pkg/service/pipeline.go [hunk 8]
```diff
@@ -1993,3 +1991,84 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+		page, pageSize, filter, orderBy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pipeline runs: %w", err)
```
Comment (jvallesm) [inline pkg/service/pipeline.go:2038]: nit: starting error messages by `failed to ...` or `couldn't ...` makes the error stack unnecessarily verbose. I prefer a description of what the code is doing as explained in [this article](https://preslav.me/2023/04/14/golang-error-handling-is-a-form-of-storytelling/#crafting-a-good-message-is-key).

File: pkg/service/pipeline.go [hunk 9]
```diff
@@ -1993,3 +1991,84 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+		requesterIDMap[pipelineRun.TriggeredBy] = struct{}{}
+	}
+
+	runnerMap := make(map[string]*string)
+	for requesterID := range requesterIDMap {
+		runner, err := s.mgmtPrivateServiceClient.CheckNamespaceByUIDAdmin(ctx, &mgmtpb.CheckNamespaceByUIDAdminRequest{Uid: requesterID})
+		if err != nil {
+			return nil, err
+		}
+		log.Info("CheckNamespaceByUIDAdmin finished", zap.String("RequesterUID", requesterID), zap.String("runnerId", runner.Id))
```
Comment (jvallesm) [inline pkg/service/pipeline.go:2052]: ❓ Is this a debug log? If not, what's its purpose?

File: pkg/repository/repository_test.go [hunk 10]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/go-redis/redismock/v9"
 	"github.com/gofrs/uuid"
 	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/require"
```
Comment (jvallesm) [inline pkg/repository/repository_test.go:19]: The whole package is tested with `quicktest`, please align with that (`require.NoError(t, err)` -> `c.Assert(err, qt.IsNil)`) if possible.

Comment (joremysh) [reply pkg/repository/repository_test.go:19]: You have also used testify here.
pkg/component/operator/base64/v0/encoding_test.go
https://github.com/instill-ai/pipeline-backend/blob/main/pkg/component/operator/base64/v0/encoding_test.go#L24

File: pkg/service/pipeline.go (hunk 9 above)
Comment (joremysh) [reply pkg/service/pipeline.go:2052]: removed

File: pkg/service/pipeline.go (hunk 8 above)
Comment (joremysh) [reply pkg/service/pipeline.go:2038]: updated

File: pkg/service/pipeline.go (hunk 6 above)
Comment (joremysh) [reply pkg/service/pipeline.go:2032]: updated

File: pkg/repository/repository_test.go (hunk 10 above)
Comment (joremysh) [reply pkg/repository/repository_test.go]: updated

File: pkg/service/pipeline.go (hunk 7 above)
Comment (joremysh) [reply pkg/service/pipeline.go]: updated

File: pkg/repository/repository_test.go (hunk 10 above)
Comment (jvallesm) [reply pkg/repository/repository_test.go]: Yes, but:
- I was talking about the `repository` package. We're importing 2 different test packages into the same file. We should avoid that first, then aim to have a single test package in `go.mod`.
- Your reference comes from a package that was in a different repository until the last release.
- Within the `component` repository we had some packages using `testify`, which predate the adoption of `quicktest`. That's the convention now but we didn't refactor the existing coverage. Using `testify` is valid in the packages that already use it (though, if these tests need to be modified, it is desirable to allocate some time to migrate to `qt`).

File: pkg/service/pipeline.go [hunk 11]
```diff
@@ -1953,3 +1954,91 @@ func (s *service) ListComponentRuns(ctx context.Context, req *pipelinepb.ListCom
...
+	if req.GetStart().IsValid() {
+		startedTimeBegin = req.GetStart().AsTime()
+	}
+	startedTimeEnd := now
+	if req.GetStop().IsValid() {
+		startedTimeEnd = req.GetStop().AsTime()
+	}
+
+	if startedTimeBegin.After(startedTimeEnd) {
+		log.Error("time range end time is earlier than start time", zap.Time("startedTimeEnd", startedTimeEnd), zap.Time("startedTimeBegin", startedTimeBegin))
```
[the remark this answers is not in the document]
Comment (joremysh) [reply pkg/service/pipeline.go]: removed the log

File: pkg/repository/repository_test.go (hunk 10 above)
Comment (joremysh) [reply pkg/repository/repository_test.go]: I see. At first I did not learn that quicktest has the checker HaveLen and IsTrue so I used testify, which I am more familiar and I think it's cleaner to write condition checks with it

**Review by jvallesm — APPROVED** [review]

File: pkg/handler/pipeline.go [hunk 12]
```diff
@@ -1994,3 +1994,20 @@ func (h *PublicHandler) ListComponentRuns(ctx context.Context, req *pb.ListCompo
 
 	return resp, nil
 }
+
+// todo: rename function to ListPipelineRunsByRequester in protobuf and update message names here
+func (h *PublicHandler) ListPipelineRunsByCreditOwner(ctx context.Context, req *pb.ListPipelineRunsByCreditOwnerRequest) (*pb.ListPipelineRunsByCreditOwnerResponse, error) {
```
Comment (jvallesm) [inline pkg/handler/pipeline.go:1999]: 👀 I guess the method rename to `ByRequester` (or, ideally, `Search`) will be done in a separate PR @joremysh ?
